EIPS/eip-5164.md

```diff
@@ -0,0 +1,211 @@
+
+## Motivation
+
+Many Ethereum protocols need to coordinate state changes across multiple EVM-based blockchains. These chains often have native or third-party bridges that allow Ethereum contracts to execute code. However, bridges have different APIs so bridge integrations are bespoke.
+
+Bridge technology is improving rapidly, and there are multiple bridges to choose from. Each one affords different properties; with varying degrees of security, speed, and control.
+
+By standardizing a cross-chain execution interface, we can cleanly separate the transport layer from the application layer. Cross-chain execution becomes another composable piece with which the Ethereum ecosystem can build shared infrastructure.
+
+## Specification
+
+This specification allows contracts on one chain to send messages to contracts on another chain. There are two key contracts:
+
+- CrossChainRelayer
+- CrossChainReceiver
+
+The `CrossChainRelayer` lives on the origin chain. Users can share a single `CrossChainRelayer` or deploy their own.
+
+The `CrossChainReceiver` lives on the destination chain and receives relayed calls.
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### CrossChainRelayer
+
+The `CrossChainRelayer` lives on the chain from which messages are sent. The Relayer's job is to broadcast the messages through a transport layer.
+
+#### Methods
+
+**relayCalls**
+
+MUST emit the `Relayed` event when successfully called.
+
+MUST increment a `nonce` so that each batch of calls can be uniquely identified.
+
+MAY require payment. Some bridges may require payment in the native currency, so the function is payable.
+
+Implementations MUST relay the Calls by calling the `receiveCalls` function on the `CrossChainReceiver` on the receiving chain with the call data.
+
+```solidity
+struct Call {
+  address target;
+  bytes data;
+  uint value;
+}
+
+interface CrossChainRelayer {
+    function relayCalls(CrossChainReceiver receiver, Call[] calldata calls, uint256 gas) external payable;
+}
+```
+
+```yaml
+- name: relayCalls
+  type: function
+  stateMutability: payable
+  inputs: 
+    - name: receiver
+      type: CrossChainReceiver
+    - name: calls
+      type: Call[]
+    - name: gas
+      type: uint256
+```
+
+#### Events
+
+**RelayedCalls**
+
+MUST be emitted by the `CrossChainRelayer` when `relayCalls` is called.
+
+```solidity
+interface CrossChainRelayer {
+  event RelayedCalls(
+      uint256 indexed nonce,
+      address indexed sender,
+      CrossChainReceiver indexed receiver,
+      Call[] calls,
+      uint256 gas
```

Comment:
Is `gas` here the total amount of gas consumed by the `calls` or is it the gas or gasLimit per call?
If this is per call, it should be an array: `uint256[] gas`